oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2TokenExchangeAuthenticationProvider.java

```diff
@@ -0,0 +1,313 @@
+		if (!subjectToken.isActive()) {
+			// As per https://tools.ietf.org/html/rfc6749#section-5.2
+			// invalid_grant: The provided authorization grant (e.g., authorization code,
+			// resource owner credentials) or refresh token is invalid, expired, revoked [...].
+			throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+		}
+
+		Authentication subjectPrincipal = subjectAuthorization.getAttribute(Principal.class.getName());
+		if (subjectPrincipal == null) {
+			// As per https://datatracker.ietf.org/doc/html/rfc8693#section-1.1
+			// We require a principal to be available via the subject_token for
+			// impersonation or delegation use cases.
+			throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+		}
+
+		OAuth2Authorization actorAuthorization = null;
+		Authentication actorPrincipal = null;
+		if (StringUtils.hasText(tokenExchangeAuthentication.getActorToken())) {
+			actorAuthorization = this.authorizationService.findByToken(
+					tokenExchangeAuthentication.getActorToken(), OAuth2TokenType.ACCESS_TOKEN);
+			if (actorAuthorization == null) {
+				throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+			}
+
+			if (this.logger.isTraceEnabled()) {
+				this.logger.trace("Retrieved authorization with actor token");
+			}
+
+			OAuth2Authorization.Token<OAuth2Token> actorToken = actorAuthorization.getToken(
+					tokenExchangeAuthentication.getActorToken());
+			if (!actorToken.isActive()) {
+				// As per https://tools.ietf.org/html/rfc6749#section-5.2
+				// invalid_grant: The provided authorization grant (e.g., authorization code,
+				// resource owner credentials) or refresh token is invalid, expired, revoked [...].
+				throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+			}
+
+			actorPrincipal = actorAuthorization.getAttribute(Principal.class.getName());
+			if (actorPrincipal == null) {
+				// As per https://datatracker.ietf.org/doc/html/rfc8693#section-1.1
+				// We require a principal to be available via the actor_token for
+				// the delegation use case.
+				throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+			}
+		}
+
+		Set<String> authorizedScopes = Collections.emptySet();
+		if (!CollectionUtils.isEmpty(tokenExchangeAuthentication.getScopes())) {
+			for (String requestedScope : tokenExchangeAuthentication.getScopes()) {
+				if (!registeredClient.getScopes().contains(requestedScope)) {
+					throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_SCOPE);
+				}
+			}
+			authorizedScopes = new LinkedHashSet<>(tokenExchangeAuthentication.getScopes());
+		}
+
+		if (this.logger.isTraceEnabled()) {
+			this.logger.trace("Validated token request parameters");
+		}
+
+		if (CollectionUtils.isEmpty(authorizedScopes)) {
+			// TODO: Validate that original authorizedScopes are also configured for this client?
+			authorizedScopes = new LinkedHashSet<>(subjectAuthorization.getAuthorizedScopes());
+		}
```

Comment:
Scopes are optional in the request. In A.1 the scopes are automatically carried over from `subject_token` to issued `access_token`. Do we need to validate that the scopes are allowed by the requesting client to carry them over in this case?

Reply:
> Do we need to validate that the scopes are allowed by the requesting client to carry them over in this case?

I think we do. Moreover, I'm not sure it's a good idea to carry-over the same scopes as-is as this may turn out to be a vulnerability through privilege escalation. My gut feeling is that we don't carry it over but we need to decide the most secure approach to take.

Reply:
For now, I've added validation for carried-over-scopes.